akka-stream/src/main/scala/akka/stream/io/Timeouts.scala

```diff
@@ -0,0 +1,166 @@
+   *
+   * There is a difference between this stage and having two idleTimeout Flows assembled into a BidiStage.
+   * If the timout is configured to be 1 seconds, then this stage will not fail even though there are elements flowing
+   * every second in one direction, but no elements are flowing in the other direction. I.e. this stage considers
+   * the *joint* frequencies of the elements in both directions.
+   */
+  def idleTimeoutBidi[A, B](timeout: FiniteDuration): BidiFlow[A, A, B, B, Unit] =
+    BidiFlow.wrap(new IdleTimeoutBidi[A, B](timeout))
+
+  private class InitialTimeout[T](timeout: FiniteDuration) extends SimpleLinearGraphStage[T] {
+
+    override def createLogic: GraphStageLogic = new SimpleLinearStageLogic {
+      private var initialHasPassed = false
+
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = {
+          initialHasPassed = true
+          push(out, grab(in))
+        }
+        override def onUpstreamFinish(): Unit = completeStage()
+        override def onUpstreamFailure(ex: Throwable): Unit = failStage(ex)
+      })
+
+      final override protected def onTimer(key: Any): Unit =
+        if (!initialHasPassed)
+          failStage(new TimeoutException(s"The first element has not yet passed through in $timeout."))
+
+      scheduleOnce("InitialTimeout", timeout)
+    }
+
+    override def toString = "InitialTimeoutTimer"
+  }
+
+  private class CompletionTimeout[T](timeout: FiniteDuration) extends SimpleLinearGraphStage[T] {
+
+    override def createLogic: GraphStageLogic = new SimpleLinearStageLogic {
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = push(out, grab(in))
+        override def onUpstreamFinish(): Unit = completeStage()
+        override def onUpstreamFailure(ex: Throwable): Unit = failStage(ex)
+      })
+
+      final override protected def onTimer(key: Any): Unit =
+        failStage(new TimeoutException(s"The stream has not been completed in $timeout."))
+
+      scheduleOnce("CompletionTimeoutTimer", timeout)
+    }
+
+    override def toString = "CompletionTimeout"
+  }
+
+  private class IdleTimeout[T](timeout: FiniteDuration) extends SimpleLinearGraphStage[T] {
+    private var nextDeadline: Deadline = Deadline.now + timeout
+
+    override def createLogic: GraphStageLogic = new SimpleLinearStageLogic {
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = {
+          nextDeadline = Deadline.now + timeout
+          push(out, grab(in))
+        }
+        override def onUpstreamFinish(): Unit = completeStage()
+        override def onUpstreamFailure(ex: Throwable): Unit = failStage(ex)
+      })
+
+      final override protected def onTimer(key: Any): Unit =
+        if (nextDeadline.isOverdue())
+          failStage(new TimeoutException(s"No elements passed in the last $timeout."))
+
+      schedulePeriodically("IdleTimeoutCheckTimer", interval = timeout / 8)
```

Comment:
do we need to do it that often? perhaps something like

    min(max(timeout / 8, 100.millis), timeout / 2)